### ISSUE TITLE
sbt 0.13 ivy resolution issue

### ISSUE THREAD

**Issue by TheSultan** [sbt/sbt#987]
## steps

`build.sbt`:

``` scala
ivyPaths :=  new IvyPaths((baseDirectory in ThisBuild).value, Some((baseDirectory in ThisBuild).value / "ivy-cache"))

name := "sbt-ivy-publicationsbug"

version := "1.0"

scalaVersion := "2.10.2"

libraryDependencies ++= Seq("org.apache.cassandra" % "cassandra-thrift" % "1.0.0" intransitive(),
                    "org.apache.thrift" % "libthrift" % "0.6.1" intransitive())
```
## original report by @TheSultan

See https://github.com/TheSultan/sbtivypublicationsbug for repro.

I believe the issue is that the maven central version of libthrift does not have a publications tag in its ivy.xml (I was able to fix just by adding that). However, I suspect that sbt may be being stricter than ivy itself and/or interpreting that flag differently, hence posting as a bug. 

Sorry, I did not debug further, since this already ate up enough of my time. But hit me back with questions if you have trouble debugging.

**Comment by copumpkin** [comment]
This triggers https://issues.scala-lang.org/browse/SI-2464 by not passing the libthrift jar into scalac.

**Comment by joelthompson** [comment]
Bump.

I also just ran into this issue. It's really subtle, seems like an extremely difficult issue for users to debug. Not sure if it's a bug in sbt or in the original maven repo, but even if it were strictly a problem with the upstream artifact, sbt should be able to handle pathological input and give users a better error message (if not just doing the right thing!).

**Comment by copumpkin** [comment]
_bump_ this has bitten us again

**Comment by eed3si9n** [comment]
We're categorizing this as Bug to take at look at this again.

**Comment by copumpkin** [comment]
Thanks!

**Comment by jsuereth** [comment]
Hmm, the reproduction wasn't working for me...

**Comment by jsuereth** [comment]
Can you guide me with some more info of what you're observing? I'd like ot try to get the reproduction working.

**Comment by TheSultan** [comment]
I just retried and experienced this

```
sultan@Maserati-2 19:20:07> git clone https://github.com/TheSultan/sbtivypublicationsbug
Cloning into 'sbtivypublicationsbug'...
remote: Reusing existing pack: 11, done.
remote: Total 11 (delta 0), reused 0 (delta 0)
Unpacking objects: 100% (11/11), done.
Checking connectivity... done
sultan@Maserati-2 19:20:17> cd sbtivypublicationsbug/
~/src/sbtivypublicationsbug ~/src ~
sultan@Maserati-2 19:20:21> ls
build.sbt  src/
sultan@Maserati-2 19:20:21> sbt
[info] Set current project to sbt-ivy-publicationsbug (in build file:/Users/sultan/src/sbtivypublicationsbug/)
> compile
[info] Updating {file:/Users/sultan/src/sbtivypublicationsbug/}sbtivypublicationsbug...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Compiling 1 Scala source to /Users/sultan/src/sbtivypublicationsbug/target/scala-2.10/classes...
[warn] Class org.apache.thrift.TBase not found - continuing with a stub.
[error] error while loading TokenRange, class file '/Users/sultan/.ivy2/cache/org.apache.cassandra/cassandra-thrift/jars/cassandra-thrift-1.0.0.jar(org/apache/cassandra/thrift/TokenRange.class)' is broken
[error] (class java.lang.NullPointerException/null)
[error] /Users/sultan/src/sbtivypublicationsbug/src/main/scala/Bug.scala:5: org.apache.cassandra.thrift.TokenRange does not have a constructor
[error]       val a = new TokenRange()
[error]               ^
[warn] one warning found
[error] two errors found
[error] (compile:compile) Compilation failed
[error] Total time: 2 s, completed Apr 21, 2014 7:20:44 PM
sultan@Maserati-2 19:20:49> sbt --version
sbt launcher version 0.13.0
sultan@Maserati-2 19:22:02> java -version
java version "1.6.0_65"
Java(TM) SE Runtime Environment (build 1.6.0_65-b14-462-11M4609)
Java HotSpot(TM) 64-Bit Server VM (build 20.65-b04-462, mixed mode)
```

**Comment by TheSultan** [comment]
More environment info just in case it is relevant

```
sultan@Maserati-2 19:23:27> ivy -version
Apache Ivy 2.3.0 - 20130110142753 :: http://ant.apache.org/ivy/
sultan@Maserati-2 19:23:31> uname -a
Darwin Maserati-2.local 12.5.0 Darwin Kernel Version 12.5.0: Sun Sep 29 13:33:47 PDT 2013; root:xnu-2050.48.12~1/RELEASE_X86_64 x86_64
```

**Comment by jsuereth** [comment]
I tried again on a different machine with the same results.  I'll see if maybe @eed3si9n can try out on a mac, as that may somehow be related to the issue. (Filesystems being what they are)

**Comment by TheSultan** [comment]
I am fairly certain I have experienced the same issue on windows and
@joelthompson experienced it on Linux so I'm not sure that is the issue.
You are using the same versions of SBT and java?

**Comment by TheSultan** [comment]
@jsuereth what output are you seeing? Successful compile? If so can you please see what libthrift you are pulling down and from where? I suspect you are getting a version without the publications issue in the ivy.xml somehow

**Comment by TheSultan** [comment]
@jsuereth, I confirmed that the issue is the ivy cache. Looks like the Maven repo for libthrift was updated to fix the ivy xml publication tag. I have checked in the offending ivy.xml and am able to reproduce a successful build and then break the build by doing the below. Please try the steps below (conceptually, you may have to modify the ivy cache path depending on your config...)

Note the key difference from before is that I am updating the source (git pull) to get the offending ivy.xml and putting that in my cache. sbt clean and clearing out my ivy cache for good measure in between things. 

```
sultan@Maserati-2 15:42:11> git clone https://github.com/TheSultan/sbtivypublicationsbug
Cloning into 'sbtivypublicationsbug'...
remote: Reusing existing pack: 16, done.
remote: Total 16 (delta 0), reused 0 (delta 0)
Unpacking objects: 100% (16/16), done.
Checking connectivity... done
sultan@Maserati-2 15:42:31> cd sbtivypublicationsbug/
~/src/sbtivypublicationsbug ~/src ~/src/sbtivypublicationsbug ~/src ~
sultan@Maserati-2 15:42:34> rm ~/.ivy2/cache/org.apache.thrift/libthrift/ivy-0.6.1.xml 
sultan@Maserati-2 15:42:56> sbt clean
[info] Set current project to sbt-ivy-publicationsbug (in build file:/Users/sultan/src/sbtivypublicationsbug/)
[success] Total time: 0 s, completed Apr 22, 2014 3:43:09 PM
sultan@Maserati-2 15:43:09> sbt compile
[info] Set current project to sbt-ivy-publicationsbug (in build file:/Users/sultan/src/sbtivypublicationsbug/)
[info] Updating {file:/Users/sultan/src/sbtivypublicationsbug/}sbtivypublicationsbug...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Compiling 1 Scala source to /Users/sultan/src/sbtivypublicationsbug/target/scala-2.10/classes...
[success] Total time: 4 s, completed Apr 22, 2014 3:43:18 PM
sultan@Maserati-2 15:43:18> cp ivy-0.6.1.xml ~/.ivy2/cache/org.apache.thrift/libthrift/ivy-0.6.1.xml 
sultan@Maserati-2 15:43:25> sbt clean
[info] Set current project to sbt-ivy-publicationsbug (in build file:/Users/sultan/src/sbtivypublicationsbug/)
[success] Total time: 0 s, completed Apr 22, 2014 3:43:31 PM
sultan@Maserati-2 15:43:31> sbt compile
[info] Set current project to sbt-ivy-publicationsbug (in build file:/Users/sultan/src/sbtivypublicationsbug/)
[info] Updating {file:/Users/sultan/src/sbtivypublicationsbug/}sbtivypublicationsbug...
[info] Resolving org.fusesource.jansi#jansi;1.4 ...
[info] Done updating.
[info] Compiling 1 Scala source to /Users/sultan/src/sbtivypublicationsbug/target/scala-2.10/classes...
[warn] Class org.apache.thrift.TBase not found - continuing with a stub.
[error] error while loading TokenRange, class file '/Users/sultan/.ivy2/cache/org.apache.cassandra/cassandra-thrift/jars/cassandra-thrift-1.0.0.jar(org/apache/cassandra/thrift/TokenRange.class)' is broken
[error] (class java.lang.NullPointerException/null)
[error] /Users/sultan/src/sbtivypublicationsbug/src/main/scala/Bug.scala:5: org.apache.cassandra.thrift.TokenRange does not have a constructor
[error]       val a = new TokenRange()
[error]               ^
[warn] one warning found
[error] two errors found
[error] (compile:compile) Compilation failed
[error] Total time: 2 s, completed Apr 22, 2014 3:43:37 PM
sultan@Maserati-2 15:43:37> 
```

**Comment by TheSultan** [comment]
You may notice that the only difference between the two ivy.xml files is the missing artifact item under publications as per my original comment:

```
sultan@Maserati-2 15:46:54> diff ivy-0.6.1.xml ~/.ivy2/cache/org.apache.thrift/libthrift/ivy-0.6.1.xml
publication="20110420164008"
---
publication="20110420154008"
&lt;artifact name="libthrift" type="jar" ext="jar" conf="master"/&gt;
```

**Comment by jsuereth** [comment]
sbt 0.13.2 launcher + java 1.6 on ubuntu.  For some reason, my ivy.xml file has that publications guy in the cache.

What sbt launcher are you using (different than sbt version)

**Comment by TheSultan** [comment]
I am using the same launcher as sbt (i.e. 0.13.0). See below for details.

However, I do not believe that is the reason you cannot reproduce. I think it is precisely because your ivy.xml contains the publications entry. Did you try removing the publications entry from your ivy.xml or overwriting that ivy.xml with the one I checked in as per my previous comment? The latter is what I did to reproduce the issue (see my previous comment for exact repro steps).

```
sultan@Maserati-2 23:04:30> sbt
[info] Set current project to sbt-ivy-publicationsbug (in build file:/Users/sultan/src/sbtivypublicationsbug/)
> sbtVersion
[info] 0.13.0
> about
[info] This is sbt 0.13.0
[info] The current project is {file:/Users/sultan/src/sbtivypublicationsbug/}sbtivypublicationsbug 1.0
[info] The current project is built against Scala 2.10.2
[info] 
[info] sbt, sbt plugins, and build definitions are using Scala 2.10.2
> sultan@Maserati-2 23:05:14> which sbt
/usr/local/bin/sbt
sultan@Maserati-2 23:05:16> cat /usr/local/bin/sbt 
#!/bin/sh
test -f ~/.sbtconfig && . ~/.sbtconfig
exec java -Xmx512M ${SBT_OPTS} -jar /usr/local/Cellar/sbt/0.13.0/libexec/sbt-launch.jar "$@"
sultan@Maserati-2 23:05:22> echo $SBT_OPTS

sultan@Maserati-2 23:05:44> echo ${SBT_OPTS}

sultan@Maserati-2 23:05:53> shasum /usr/local/Cellar/sbt/0.13.0/libexec/sbt-launch.jar
b78a29b8db7ae0fc90a7890a84a4617975e1d450  /usr/local/Cellar/sbt/0.13.0/libexec/sbt-launch.jar
sultan@Maserati-2 23:06:59> 
```

**Comment by eed3si9n** [comment]
I cannot reproduce this bug. To avoid any cache contamination, I've added the following:

``` scala
ivyPaths :=  new IvyPaths((baseDirectory in ThisBuild).value, Some((baseDirectory in ThisBuild).value / "ivy-cache"))
```

This creates a fresh directory for ivy cache. Also I added `project/build.properties`:

``` scala
sbt.version=0.13.6-M2
```

The generated ivy-0.6.1.xml is uploaded [here](https://gist.github.com/eed3si9n/26ebc4aca0b7b6f3ff44):

``` xml
    <info organisation="org.apache.thrift"
        module="libthrift"
        revision="0.6.1"
        status="release"
        publication="20110420164008"
    >
```

compile ran successfully.

Your analysis that sbt being more strict than Ivy (or older version of Ivy) may still hold. If you're using Ivy and/or Ivy-based product, I suggest you use `-Dsbt.ivy.home` to change the Ivy paths for all builds.
